docs/specs/alcproxies.md

```diff
@@ -0,0 +1,251 @@
+
+Concerns with gRPC:
+* Encoding objects in a way that we can manually decode and pass them into the `TargetObject` needs to be investigated further, to make sure it works.
+* At the very least, there will probably be more overhead with how gRPC does in-process communication calls. At worst, it won't work since most of it is designed for use with ASP.NET instead of just .NET Core.
+* Limited to using http/2 for transport.
+
+
+The current plan is to build with either a second `DispatchProxy` for in-proc ALCs and gRPC for cross-proc connections, or use gRPC for both in- and out-of-proc work. This is since serialization has a lot of potential issues that may affect the API later down the road. gRPC will probably
+
+### Extensibility
+One nice thing about the design is its ability to have its components "slotted" in and out. The main part of the project that has multiple options/paths that could be taken are all in the implementation of `ServerObject` and `ClientObject`. As long as the client can receive an instruction with info from the `ProxyObject`, and the server can invoke the `TargetObject`, the middle communication between the two can be implemented in multiple ways. If implemented right, this allows for the project to be open for anyone to add their own implementation of the communication between ALCs to suit their needs, possibly using some of the alternative ideas listed in this document.
+
+### Notes on handling specifics
+
+#### Versioning
+There are a few guidelines we're depending on to make sure versioning works for our interfaces:
+* The name of the interface needs to remain the same.
+* Method signatures that have already been created can't be edited or deleted, though new ones can be added.
+
+There may be cases where the main program attempts to make a method call to an older plugin, where the method doesn't exist. When the API makes a call to `TargetObject` through `ServerObject`, it will be using `Reflection.Invoke`, meaning if there ever is a case where we don't have the requested method we're calling, the invoke should throw a `TargetException`, which we could *potentially* do something with, but probably not.
+
+#### ALC Unloading
+During the `AssemblyLoadContext.Unloading` event, the `ClientObject` can remove any reference it has to the `ServerObject`, allowing for the server to unload with the ALC. Since there are no other connections into the ALC other then the connection between client and server, the target ALC should be able to unload everything in itself correctly. Afterwards, the `ProxyObject` could remove its reference to the `ClientObject`, opening it up to be collected by the GC. 
+
+There's no need for the server to dereference the target object, since they are both within the target ALC, and can be removed during the unload process.
+
+There are a few potential places where this isolation isn't perfect, which is discussed in the next point.
```

Comment:
As mentioned above this is a bad choice. We shouldn't allow unloading while there are client references.

Reply:
> We shouldn't allow unloading while there are client references.

This is an interesting debate.  On one hand, allowing the requested unload to be more predictable and the proxy to throw when the object is no longer alive may be desirable (this would be my vote).  On the other hand, keeping the reference alive until all references are done is the current behavior and leads to unpredictable unloads.  The value add of having proxies is better diagnostics when things are unloaded (by throwing) as opposed to just not being unloaded.

Reply:
See my comment below. If we support this we need to support both options.

While any thread is executing code in the ALC, it must not be unloaded. I suspect the existing unloadability design already prevents this completely. So we may get the best of both worlds.

It still will not evict a running thread.

Reply:
> better diagnostics when things are unloaded (by throwing)

It is debatable.  Coming form a `C++` development background I tend to agree with you, but it is not really consistent with GC approach of `C#`

Reply:
If we really want to support forced unload, we would probably need a feature in ALC which would prevent holding a hard reference to an object in the ALC which supported forced unload.

This may simply provide another mechanism to hold references safely

Reply:
>See my comment below. If we support this we need to support both options.

This is a bit hard to do consistently. As long as out of process scenarios are an option proxies must understand the concept of unloadability since by definition the called shouldn't care whether it is out of proc or in proc.. I agree in principle that in proc servers will probably not be unloadable or will be handled by the GC, but to the consumer of the proxy that shouldn't matter. The server can always be disconnected and that must be accepted by the client as long as out of proc scenarios are possible.

Reply:
Proxies alone will not make unloadability 100% predictable, but they will help a ton. Especially if they react to unload event and "eagerly" cut the references. That would also help with diagnostics why it didn't unload as there will be a lot less GC references to go through.

It does not solve the problem of threads running in the ALC, but that is something we should not try to fix - that leads to the "rude abort" issues we've had in all our runtimes which is a sad story.

Reply:
@AaronRobinsonMSFT 

> both options

* Immediately terminate the server when any thread call `AssemblyLoadContext.Unload()`
* Don't immediately terminate the server when any thread call `AssemblyLoadContext.Unload()`

We shouldn't try to guess the policy for the developer. Both are valid options.
We shouldn't take away the developers choices.

Reply:
>Immediately terminate the server when any thread call AssemblyLoadContext.Unload()

That isn't how that function is defined though so I don't see the concern. The documentation specifically calls out "Unloading will not occur while there are references to the AssemblyLoadContext."? Which makes rude unloading invalid unless a new API is added.

Reply:
Am I missing a proposal here for new semantics?

Reply:
Would it make sense to add a manual `Unload()` method to the proxy then (or some equivalent)? It's kind of hard to remove the reference to the object without going through the proxies, but since they don't share that much to the user (since they're trying to emulate the object itself) there may not be an easy way to get users unloading without waiting for the GC to collect the de-referenced proxy, then unload the rest of the ALC. I could also add in a similar option to ALCs `isCollectible` that would let the developer choose whether they want the automatic unload, or something more manual.

Reply:
@cablooshe From re-reading this I think Steve and I are arguing for the same thing, just coming at it from two different sides. I am looking at this from the idea that a client has little control over an out of proc server, so "unloading" needs to be resilient and Steve's point is that unilaterally permitting the unloading of an ALC in a rude manner isn't something we want nor should be permitted - which I would agree with. We should prefer letting the GC handle clean up and unload when free rather than allow the client to say "rudely unload no matter what". I was trying to state that proxies must be able to handle a severed connection and should be resilient regardless of what the client requests since out of proc servers can be rudely "unloaded" (i.e. shutdown).

I don't think changing the semantics of the `Unload()` function on the ALC to support rudely being unloaded is a good idea in general and that definitely goes for proxies since there could be an unknown number of proxies to the same server.

Reply:
We don't want to change the semantics of `Unload()` - refs to the target ALC that don't go through the proxy object will keep it alive like they do currently. However it may be useful to provide functionality along the lines of "if you only create and operate on types in an ALC using the proxy API", then we optionally sever the connection on `Unload()`. It's not providing any guarantees about forcible unloads, but may be closer to the desired semantics for some plugin models. Whether we do this or not, I think the more important part of this is for the client to be resilient to a terminated server as you've said.

Reply:
In case my point was missed, in order for the client to be resilient, the proxies need well defined behavior when exceptional behavior occurs.